Hashmap.h

```diff
@@ -0,0 +1,318 @@
+
+    iterator(HashMapData *data, size_t index, size_t position)
+            : data_(data), index_(index), position_(position) {}
+
+    bool operator==(const iterator &other) const {
+      return data_ == other.data_ && index_ == other.index_ && position_ == other.position_;
+    }
+
+    bool operator!=(const iterator &other) const {
+      return !((*this) == other);
+    }
+
+    iterator &operator++() {
+      ++position_;
+
+      if (position_ == (*data_)[index_].size()) {
+        position_ = 0;
+        ++index_;
+        while (index_ < data_->size() && (*data_)[index_].size() == 0) {
+          ++index_;
+        }
+      }
+      return *this;
+    }
+
+    const iterator operator++(int) {
+      const iterator current = (*this);
+      ++(*this);
+      return current;
+    }
+
+    HashMapItem *operator->() {
+      return ((*data_)[index_].begin() + position_).operator->();
+    }
+
+    HashMapItem &operator*() {
+      return ((*data_)[index_].begin() + position_).operator*();
+    }
+
+  private:
+    HashMapData *data_;
+    size_t index_;
+    size_t position_;
+  };
+
+  class const_iterator {
+  public:
+    const_iterator() : data_(nullptr), index_(0), position_(0) {}
+
+    const_iterator(const HashMapData *data, size_t index, size_t position)
+            : data_(data), index_(index), position_(position) {}
+
+    bool operator==(const const_iterator &other) const {
+      return data_ == other.data_ && index_ == other.index_ && position_ == other.position_;
+    }
+
+    bool operator!=(const const_iterator &other) const {
+      return !((*this) == other);
+    }
+
+    const_iterator &operator++() {
+      ++position_;
+      if (position_ == (*data_)[index_].size()) {
+        position_ = 0;
+        ++index_;
+        while (index_ < data_->size() && (*data_)[index_].size() == 0) {
+          ++index_;
+        }
+      }
+      return *this;
+    }
+
+    const const_iterator operator++(int) {
+      const const_iterator current = (*this);
+      ++(*this);
+      return current;
+    }
+
+    const HashMapItem *operator->() const {
+      return ((*data_)[index_].cbegin() + position_).operator->();
+    }
+
+    const std::pair<const KeyType, ValueType> &operator*() const {
+      return ((*data_)[index_].cbegin() + position_).operator*();
+    }
+
+  private:
+    const HashMapData *data_ = nullptr;
+    size_t index_ = 0;
+    size_t position_ = 0;
+  };
+
+  HashMap(const HashMap<KeyType, ValueType, Hash> &other) : HashMap(other.begin(), other.end(), other.hash_) {}
+
+  HashMap &operator=(const HashMap<KeyType, ValueType, Hash> &other) {
+    size_ = other.size_;
+    hash_ = other.hash_;
+    batches_ = other.batches_;
+
+    HashMapData new_data(batches_);
+
+    for (const auto item : other) {
+      auto key = item.first;
+      auto value = item.second;
+      new_data[getHash(key)].emplace_back(key, value);
+    }
+
+    std::swap(data_, new_data);
```

Comment:
https://www.geeksforgeeks.org/difference-between-stdswap-and-stdvectorswap/

Reply:
https://en.cppreference.com/w/cpp/container/vector/swap2